### PR TITLE
[appkit] add signatures to NSGridView.Create and deprecate the actual one 

### PR DESCRIPTION
**Pull request by joconte**
Fixes https://github.com/xamarin/xamarin-macios/issues/6232

I'm not sure about this one but as you guys said PR are always welcomed so I give it a try 😄

### CODE REVIEW

**Comment by joconte** [comment]
Aaahhh my commits history is not clean...

**Comment by chamons** [comment]
No worries - we'll squash commit history away on merge anyway.

File: src/appkit.cs [hunk 1]
```diff
@@ -7216,10 +7216,20 @@ interface NSGridView
 		[Static]
 		[Export ("gridViewWithNumberOfColumns:rows:")]
 		NSGridView Create (nint columnCount, nint rowCount);
-
+		
+#if XAMCORE_4_0
```
Comment (joconte) [inline src/appkit.cs]: I made a mistake I wanted to type 
```C#
#if !XAMCORE_4_0
```

**Comment by spouliot** [comment]
build

File: src/Foundation/NSArray.cs [hunk 2]
```diff
@@ -63,7 +63,33 @@ public static NSArray FromNSObjects (int count, params INativeObject [] items)
...
+			return ret;
+		}
+		public static NSArray FromNSObjects<T> (T [,] items) where T: class, INativeObject
+		{
+			if (items == null)
+				return null;
+			
+			var width = items.GetLength (0);
+			var height = items.GetLength (1);
+			var ret = new NSMutableArray ((nuint)(width * height));
```
Comment (rolfbjarne) [inline src/Foundation/NSArray.cs]: This won't do what the API you're trying to do expects, because it expects an NSArray of NSArray of NSView, not an NSArray of NSView.

I think the easiest is to convert to a `T[][]` and then call that overload.

```csharp
var width = items.GetLength (0);
var height = items.GetLength (1);
var array = new T [height][];
for (var y = 0; y < height; y++) {
    var row = new T [width];
    for (var x = 0; x < width; x++) {
        row [x] = items [x, y];
    }
    items [y] = row;
}
return FromNSObjects (items);
```


```

File: src/appkit.cs [hunk 3]
```diff
@@ -7216,10 +7216,22 @@ interface NSGridView
 		[Static]
 		[Export ("gridViewWithNumberOfColumns:rows:")]
 		NSGridView Create (nint columnCount, nint rowCount);
-
+		
+#if !XAMCORE_4_0
+		[EditorBrowsable (EditorBrowsableState.Never)]
+		[Obsolete]
```
Comment (rolfbjarne) [inline src/appkit.cs]: The Obsolete attribute needs a description (to tell the user which overload to use instead).

**Comment by joconte** [comment]
@rolfbjarne Thanks, I will fix it after work.

**Comment by rolfbjarne** [comment]
build

**Comment by joconte** [comment]
I'm probably making a mistake with the tests but I wanted to give it a try.

File: tests/xammac_tests/AppKit/NSGridViewTest.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		{
+			NSView [][] nSViewsArrayOfArray = new NSView [4][];
+			nSViewsArrayOfArray [0] = new NSView [1] { new NSView() };
+			nSViewsArrayOfArray [1] = new NSView [1] { new NSView() };
+			nSViewsArrayOfArray [2] = new NSView [1] { new NSView() };
+			nSViewsArrayOfArray [3] = new NSView [1] { new NSView() };
+            
+			NSGridView nSGridViewArrayOfArray = NSGridView.Create (nSViewsArrayOfArray);
+			
+			Assert.NotNull(nSGridViewArrayOfArray);
```
Comment (rolfbjarne) [inline tests/xammac_tests/AppKit/NSGridViewTest.cs]: This is a great start!

However, I'd add more asserts to make sure things really work as expected:

 1. Make each NSView different, doesn't matter what exactly, it's just so that you can assert you get the right one later. For instance, make them all `NSTextView`s and set the Value property:

    ```csharp
    nSViewsArrayOfArray [0] = new NSView [1] { new NSTextView ()  { Value = "0" } };
    nSViewsArrayOfArray [1] = new NSView [1] { new NSTextView ()  { Value = "1" } };
    // ...
    ```

    And then assert that each cell is the expected one:

    ```csharp
    Assert.AreEqual ("0", ((NSTextView) nSGridViewArrayOfArray.GetCell (0, 0)).Value, "0,0");
    Assert.AreEqual ("1", ((NSTextView) nSGridViewArrayOfArray.GetCell (0, 1)).Value, "0,1");
    // ...
    ```

3. also what happens with other types of input, such as arrays of different length:

    ```csharp
    nSViewsArrayOfArray [0] = new NSView [1] { new NSTextView ()  { Value = "0" } };
    nSViewsArrayOfArray [1] = new NSView [2] { new NSTextView ()  { Value = "1" }, new NSTextView () { Value = "1bis" } };
    // ...
    ```

4. or arrays with nulls:

    ```csharp
    nSViewsArrayOfArray [0] = new NSView [1] { new NSTextView ()  { Value = "0" } };
    nSViewsArrayOfArray [1] = new NSView [1] { null } };
    // ...
    ```

5. or null arrays:

    ```csharp
    nSViewsArrayOfArray [0] = new NSView [1] { new NSTextView ()  { Value = "0" } };
    nSViewsArrayOfArray [1] = null;
    // ...
    ```

6. or even just null input altogether:


    ```csharp
    nSViewsArrayOfArray = null;
    ```

Basically try to break your own code :smile:

File: tests/xammac_tests/AppKit/NSGridViewTest.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+		{
+			NSView [,] nSViewsTwoDim = new NSView [2, 2];
+			nSViewsTwoDim [0, 0] = new NSView ();
+			nSViewsTwoDim [0, 1] = new NSView ();
+			nSViewsTwoDim [1, 0] = new NSView ();
+			nSViewsTwoDim [1, 1] = new NSView ();
+
+			NSGridView nSGridViewTwoDimensionalArray = NSGridView.Create (nSViewsTwoDim);
+			
+			Assert.NotNull(nSGridViewTwoDimensionalArray);
```
Comment (rolfbjarne) [inline tests/xammac_tests/AppKit/NSGridViewTest.cs]: Same here: be create about the input, and when valid input, add asserts for each cell (in particular this makes sure the x/y axis aren't mixed up).

File: tests/xammac_tests/xammac_tests.csproj [hunk 6]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (rolfbjarne) [inline tests/xammac_tests/xammac_tests.csproj]: Unnecessary whitespace change.

File: tests/xammac_tests/xammac_tests.csproj [hunk 7]
```diff
@@ -190,7 +191,7 @@
   </ItemGroup>
   <ItemGroup>
     <Folder Include="EmbeddedResources\" />
-	<Folder Include="..\monotouch-test\BundleResourceResources\Base.lproj\" />
+    <Folder Include="..\monotouch-test\BundleResourceResources\Base.lproj\" />
```
Comment (rolfbjarne) [inline tests/xammac_tests/xammac_tests.csproj]: Same: unnecessary whitespace change.

File: tests/xammac_tests/xammac_tests.csproj [hunk 8]
```diff
@@ -206,7 +207,7 @@
     <BundleResource Include="..\monotouch-test\CoreImage\xamarinmonkey.heic">
       <Link>CoreImage\xamarinmonkey.heic</Link>
     </BundleResource>
-	<BundleResource Include="..\monotouch-test\Resources\Base.lproj\Localizable.strings" >
+    <BundleResource Include="..\monotouch-test\Resources\Base.lproj\Localizable.strings">
```
Comment (rolfbjarne) [inline tests/xammac_tests/xammac_tests.csproj]: Same: unnecessary whitespace change.

**Comment by joconte** [comment]
@rolfbjarne Thanks for the help, I will fix this after school.

File: tests/xammac_tests/xammac_tests.csproj (hunk 8 above)
Comment (joconte) [reply tests/xammac_tests/xammac_tests.csproj]: fixed

File: tests/xammac_tests/xammac_tests.csproj (hunk 7 above)
Comment (joconte) [reply tests/xammac_tests/xammac_tests.csproj]: fixed

File: tests/xammac_tests/xammac_tests.csproj (hunk 6 above)
Comment (joconte) [reply tests/xammac_tests/xammac_tests.csproj]: fixed

**Comment by joconte** [comment]
I pushed some more tests but I have problems with the ```null```suggestions

```C#
[Test]
        public void CreateWithNSViewArrayOfArrayCheckWithNullArray()
        {
            NSView [][] nSViewsArrayOfArray = new NSView [3][];
            nSViewsArrayOfArray [0] = new NSView [1] { new NSTextView ()  { Value = "0" } };
            nSViewsArrayOfArray [1] = null;
            NSGridView nSGridViewArrayOfArray = NSGridView.Create (nSViewsArrayOfArray);

            Assert.NotNull (nSGridViewArrayOfArray);

            Assert.AreEqual ("0", ((NSTextView)(nSGridViewArrayOfArray.GetCell (0, 0)).ContentView).Value, "0,0");
            Assert.AreEqual (null,nSGridViewArrayOfArray.GetRow (1)); //Crash
        }
```

Same problem here : 
```C#
[Test]
        public void CreateWithNSViewArrayOfArrayCheckWithNull()
        {
            NSView [][] nSViewsArrayOfArray = new NSView [2][];
            nSViewsArrayOfArray [0] = new NSView [1] { new NSTextView ()  { Value = "0" } };
            nSViewsArrayOfArray [1] = new NSView [1] { null } ; //crash here

            NSGridView nSGridViewArrayOfArray = NSGridView.Create (nSViewsArrayOfArray);

            Assert.NotNull (nSGridViewArrayOfArray);
            Assert.AreEqual ("0", ((NSTextView)(nSGridViewArrayOfArray.GetCell (0, 0)).ContentView).Value, "0,0");
            Assert.AreEqual (null, ((nSGridViewArrayOfArray.GetCell (0, 1))), "0,1"); //CRASH
        }
```

I was wondering, is it supposed to be allowed on the native ```(instancetype)gridViewWithViews:(NSArray<NSArray<NSView *> *> *)rows;```?

**Comment by rolfbjarne** [comment]
@joconte how does it crash? can you share the application output?

**Comment by joconte** [comment]
@rolfbjarne 
here you go : 
https://gist.github.com/joconte/632657bbcdb9fe90cc6123b3c46ff6b5
https://gist.github.com/joconte/9915af805ec9694f4c2185643e9ea5bb

**Comment by rolfbjarne** [comment]
@joconte yes, so it looks like the native API doesn't work with that input. This means you have to validate the input in the methods you implemented (and throw an ArgumentOutOfRangeException/ArgumentNullException), and in the tests assert that you get those exceptions:

```csharp
Assert.Throws<ArgumentNullException> (() => NSGridView.Create (brokenArray), "Broken Array #1");
```

File: src/Foundation/NSArray.cs [hunk 9]
```diff
@@ -63,7 +63,44 @@ public static NSArray FromNSObjects (int count, params INativeObject [] items)
...
+				return null;
+			
+			var width = items.GetLength (0);
+			var height = items.GetLength (1);
+			var ret = new T [height][];
+			for (var y = 0; y < height; y++) {
+				var row = new T [width];
+				for (var x = 0; x < width; x++) {
+					//if (items [x, y] == null)
+					//	throw new ArgumentNullException (); not sure if I should catch the error here
```
Comment (joconte) [inline src/Foundation/NSArray.cs]: I need an advice here

File: src/Foundation/NSArray.cs [hunk 10]
```diff
@@ -63,7 +63,44 @@ public static NSArray FromNSObjects (int count, params INativeObject [] items)
...
 		}
+		public static NSArray FromNSObjects<T> (params T [][] items) where T: class, INativeObject
+		{
+			if (items == null)
+				return null;
+			
+			var ret = new NSMutableArray ((nuint)items.Length);
+			foreach (var t in items) {
+				if (t == null)
+					throw new ArgumentNullException(nameof (t));
```
Comment (joconte) [inline src/Foundation/NSArray.cs]: Is the use of ```C# nameof()```correct here ?

Comment (rolfbjarne) [reply src/Foundation/NSArray.cs]: No, the caller does not know about local variables, so it does not make sense to use it in the exception message.

I would do something like this:

```csharp
for (var i = 0; i < items.Length; i++) {
    var row = items [i];
    if (row == null)
        throw new ArgumentNullException (nameof (items), $"Element [{i}] is null");
    for (var j = 0; j < row.Length; j++) {
        var element = row [j];
        if (element == null)
            throw new ArgumentNullException (nameof (items), $"Element [{i}][{j}] is null");
       ret.Add (NSArray.FromNSObjects (row));
    }
}
```

File: src/Foundation/NSArray.cs (hunk 9 above)
Comment (rolfbjarne) [reply src/Foundation/NSArray.cs]: I don't think you need it as long as the other overload throws the correct exception.

File: tests/xammac_tests/AppKit/NSGridViewTest.cs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+		{
+			NSView [][] nSViewsArrayOfArray = new NSView [4][];
+			nSViewsArrayOfArray [0] = new NSView [1] { new NSTextView ()  { Value = "0" } };
+			nSViewsArrayOfArray [1] = new NSView [1] { new NSTextView ()  { Value = "1" } };
+			nSViewsArrayOfArray [2] = new NSView [1] { new NSTextView ()  { Value = "2" } };
+			nSViewsArrayOfArray [3] = new NSView [1] { new NSTextView ()  { Value = "3" } };
+			
+			NSGridView nSGridViewArrayOfArray = NSGridView.Create (nSViewsArrayOfArray);
+
+			Assert.NotNull(nSGridViewArrayOfArray);
```
Comment (rolfbjarne) [inline tests/xammac_tests/AppKit/NSGridViewTest.cs]:
```suggestion
			Assert.NotNull (nSGridViewArrayOfArray);
```

File: tests/xammac_tests/AppKit/NSGridViewTest.cs [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+			nSViewsTwoDim [0, 1] = new NSTextView() {Value = "1"};
+			nSViewsTwoDim [1, 0] = new NSTextView() {Value = "2"};
+			nSViewsTwoDim [1, 1] = new NSTextView() {Value = "3"};
+
+			NSGridView nSGridViewTwoDimensionalArray = NSGridView.Create (nSViewsTwoDim);
+			
+			Assert.NotNull (nSGridViewTwoDimensionalArray);
+			Assert.AreEqual ("0", ((NSTextView)(nSGridViewTwoDimensionalArray.GetCell (0, 0)).ContentView).Value, "0,0");
+			Assert.AreEqual ("1", ((NSTextView)(nSGridViewTwoDimensionalArray.GetCell (0, 1)).ContentView).Value, "0,1");
+			Assert.AreEqual ("2", ((NSTextView)(nSGridViewTwoDimensionalArray.GetCell (1, 0)).ContentView).Value, "0,2");
```
Comment (rolfbjarne) [inline tests/xammac_tests/AppKit/NSGridViewTest.cs]:
```suggestion
			Assert.AreEqual ("2", ((NSTextView)(nSGridViewTwoDimensionalArray.GetCell (1, 0)).ContentView).Value, "1,0");
```

File: tests/xammac_tests/AppKit/NSGridViewTest.cs [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+			Assert.AreEqual ("3", ((NSTextView)(nSGridViewTwoDimensionalArray.GetCell (1, 1)).ContentView).Value, "0,3");
```
Comment (rolfbjarne) [inline tests/xammac_tests/AppKit/NSGridViewTest.cs]:
```suggestion
			Assert.AreEqual ("3", ((NSTextView)(nSGridViewTwoDimensionalArray.GetCell (1, 1)).ContentView).Value, "1,1");
```

**Comment by rolfbjarne** [comment]
build
